lib/pavilion/test_config/resolver.py

```diff
@@ -0,0 +1,933 @@
+
+    @staticmethod
+    def resolve_permutations(test_cfg, base_var_man):
+        """Resolve permutations for all used permutation variables, returning a
+        variable manager for each permuted version of the test config. We use
+        this opportunity to populate the variable manager with most other
+        variable types as well.
+
+        :param dict test_cfg: The raw test configuration dictionary.
+        :param variables.VariableSetManager base_var_man: The variables for
+            this config (absent the scheduler variables).
+        :returns: The modified configuration, and a list of variable set
+            managers, one for each permutation.
+        :rtype: (dict, [variables.VariableSetManager])
+        :raises TestConfigError: When there are problems with variables or the
+            permutations.
+        """
+
+        permute_on = test_cfg['permute_on']
+        del test_cfg['permute_on']
+
+        used_per_vars = set()
+        for per_var in permute_on:
+            try:
+                var_set, var, index, subvar = base_var_man.resolve_key(per_var)
+            except KeyError:
+                raise TestConfigError(
+                    "Permutation variable '{}' is not defined."
+                    .format(per_var))
+            if index is not None or subvar is not None:
+                raise TestConfigError(
+                    "Permutation variable '{}' contains index or subvar."
+                    .format(per_var))
+            elif base_var_man.is_deferred(var_set, var):
+                raise TestConfigError(
+                    "Permutation variable '{}' references a deferred variable."
+                    .format(per_var))
+            used_per_vars.add((var_set, var))
+
+        # var_men is a list of variable managers, one for each permutation
+        var_men = base_var_man.get_permutations(used_per_vars)
+        for var_man in var_men:
+            var_man.resolve_references(string_parser.parse)
+        return test_cfg, var_men
+
+    NOT_OVERRIDABLE = ['name', 'suite', 'suite_path', 'scheduler']
+
+    def apply_overrides(self, test_cfg, overrides):
+        """Apply overrides to this test.
+
+        :param dict test_cfg: The test configuration.
+        :param list overrides: A list of raw overrides in a.b.c=value form.
+        :raises: ValueError, KeyError
+    """
+
+        for ovr in overrides:
+            if '=' not in ovr:
+                raise ValueError(
+                    "Invalid override value. Must be in the form: "
+                    "<key>=<value>. Ex. -c run.modules=['gcc'] ")
+
+            key, value = ovr.split('=', 1)
+            key = key.split('.')
+
+            self._apply_override(test_cfg, key, value)
+
+        TestConfigLoader().validate(test_cfg)
+
+    def _apply_override(self, test_cfg, key, value):
+        """Set the given key to the given value in test_cfg.
+        :param dict test_cfg: The test configuration.
+        :param [str] key: A
```

Comment:
And... you seem to have trailed off here.  Also, specify return type and possible exception types.